### ISSUE TITLE
NoSuchFileException is thrown when trace file path does not exist

### ISSUE THREAD

**Issue by ewels** [nextflow-io/nextflow#334]
It appears that trying to use dynamic filenames in the config for timeline and trace files doesn't work. For example:

```groovy
timeline {
  enabled = true
  file = "results/NGI-RNAseq_timeline.html"
}
trace {
  enabled = true
  file = "results/NGI-RNAseq_trace.txt"
}
```

produces output files as expected. However, if I specify:

```groovy
timeline {
  enabled = true
  file = { "${params.outdir}/NGI-RNAseq_timeline.html" }
}
trace {
  enabled = true
  file = { "${params.outdir}/NGI-RNAseq_trace.txt" }
}
```

Then I get two files in the current directory called `testing$_run_closure4$_closure7@740d2e78` and `testing$_run_closure5$_closure8@6b88ca8c`. The contents of these files seems to be fine.

Phil

**Comment by pditommaso** [comment]
You can use variables without the need of a closure.

**Comment by ewels** [comment]
Ok, but then I get a different error..

```groovy
timeline {
  enabled = true
  file = "${params.outdir}/NGI-RNAseq_timeline.html"
}
trace {
  enabled = true
  file = "${params.outdir}/NGI-RNAseq_trace.txt"
}
```

Then:
```
$ nextflow run ../main.nf -resume -profile testing --gtf ./test_data/ngi-rna_test_set/genes.gtf --bed12 ./test_data/ngi-rna_test_set/genes.bed --hisat2_index ./test_data/ngi-rna_test_set/r64/ --aligner hisat2 --reads "./test_data/ngi-rna_test_set/*.fastq.gz"
-----
N E X T F L O W  ~  version 0.24.0
Launching `../main.nf` [boring_meitner] - revision: 30f5a0f573
ERROR ~ No such file: /Users/philewels/GitHub/NGI-RNAseq/tests/[:]/NGI-RNAseq_trace.txt

 -- Check '.nextflow.log' file for details
```

**Comment by pditommaso** [comment]
You need to define that parameter in the config file before referencing it.

**Comment by ewels** [comment]
Ok - it's already defined in the main pipeline script. But if I add it to the config file as well then I get a slightly different error:

```
params {
  outdir = './results'
}
timeline {
  enabled = true
  file = "${params.outdir}/NGI-RNAseq_timeline.html"
}
trace {
  enabled = true
  file = "${params.outdir}/NGI-RNAseq_trace.txt"
}
```
Gives:
```
nextflow run ../main.nf -resume -profile testing --gtf ./test_data/ngi-rna_test_set/genes.gtf --bed12 ./test_data/ngi-rna_test_set/genes.bed --hisat2_index ./test_data/ngi-rna_test_set/r64/ --aligner hisat2 --reads "./test_data/ngi-rna_test_set/*.fastq.gz"
-----
N E X T F L O W  ~  version 0.24.0
Launching `../main.nf` [high_swirles] - revision: 30f5a0f573
ERROR ~ No such file: /Users/philewels/GitHub/NGI-RNAseq/tests/results/NGI-RNAseq_trace.txt

 -- Check '.nextflow.log' file for details
```
Full log:
```
May-05 10:18:48.176 [main] DEBUG nextflow.cli.Launcher - $> /usr/local/bin/nextflow run ../main.nf -resume -profile testing --gtf ./test_data/ngi-rna_test_set/genes.gtf --bed12 ./test_data/ngi-rna_test_set/genes.bed --hisat2_index ./test_data/ngi-rna_test_set/r64/ --aligner hisat2 --reads ./test_data/ngi-rna_test_set/*.fastq.gz
May-05 10:18:48.461 [main] INFO  nextflow.cli.CmdRun - N E X T F L O W  ~  version 0.24.0
May-05 10:18:48.497 [main] INFO  nextflow.cli.CmdRun - Launching `../main.nf` [high_swirles] - revision: 30f5a0f573
May-05 10:18:48.534 [main] DEBUG nextflow.config.ConfigBuilder - Found config home: /Users/philewels/.nextflow/config
May-05 10:18:48.535 [main] DEBUG nextflow.config.ConfigBuilder - Found config base: /Users/philewels/GitHub/NGI-RNAseq/nextflow.config
May-05 10:18:48.544 [main] DEBUG nextflow.config.ConfigBuilder - Parsing config file: /Users/philewels/.nextflow/config
May-05 10:18:48.548 [main] DEBUG nextflow.config.ConfigBuilder - Parsing config file: /Users/philewels/GitHub/NGI-RNAseq/nextflow.config
May-05 10:18:49.043 [main] DEBUG nextflow.config.ConfigBuilder - Setting config profile: 'testing'
May-05 10:18:49.334 [main] DEBUG nextflow.Session - Session uuid: 39c6f3fa-0ce1-4189-8f1e-992521e1a362
May-05 10:18:49.334 [main] DEBUG nextflow.Session - Run name: high_swirles
May-05 10:18:49.336 [main] DEBUG nextflow.Session - Executor pool size: 4
May-05 10:18:49.367 [main] DEBUG nextflow.cli.CmdRun -
  Version: 0.24.0 build 4235
  Modified: 20-03-2017 08:39 UTC (09:39 CEST)
  System: Mac OS X 10.12.4
  Runtime: Groovy 2.4.10 on Java HotSpot(TM) 64-Bit Server VM 1.8.0_25-b17
  Encoding: UTF-8 (UTF-8)
  Process: 50593@PhilEMBA.scilifelab.se [130.229.42.149]
  CPUs: 4 - Mem: 8 GB (163.1 MB) - Swap: 5 GB (1.3 GB)
May-05 10:18:49.412 [main] DEBUG nextflow.Session - Work-dir: /Users/philewels/GitHub/NGI-RNAseq/tests/work [Mac OS X]
May-05 10:18:49.609 [main] DEBUG nextflow.Session - Session start invoked
May-05 10:18:49.616 [main] DEBUG nextflow.processor.TaskDispatcher - Dispatcher > start
May-05 10:18:49.617 [main] DEBUG nextflow.trace.TraceFileObserver - Flow starting -- trace file: /Users/philewels/GitHub/NGI-RNAseq/tests/results/NGI-RNAseq_trace.txt
May-05 10:18:49.754 [main] ERROR nextflow.cli.Launcher - /Users/philewels/GitHub/NGI-RNAseq/tests/results/NGI-RNAseq_trace.txt
java.nio.file.NoSuchFileException: /Users/philewels/GitHub/NGI-RNAseq/tests/results/NGI-RNAseq_trace.txt
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
	at java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:434)
	at java.nio.file.Files.newOutputStream(Files.java:216)
	at java.nio.file.Files.newBufferedWriter(Files.java:2860)
	at nextflow.trace.TraceFileObserver.onFlowStart(TraceFileObserver.groovy:202)
	at nextflow.Session$_start_closure12.doCall(Session.groovy:405)
	at nextflow.Session$_start_closure12.call(Session.groovy)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2030)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2015)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2044)
	at nextflow.Session.start(Session.groovy:405)
	at nextflow.script.ScriptRunner.execute(ScriptRunner.groovy:151)
	at nextflow.cli.CmdRun.run(CmdRun.groovy:223)
	at nextflow.cli.Launcher.run(Launcher.groovy:410)
	at nextflow.cli.Launcher.main(Launcher.groovy:564)
```

**Comment by pditommaso** [comment]
If you specify the same file statically, does it work?

**Comment by ewels** [comment]
Yup, see first issue post above..

**Comment by ewels** [comment]
Aha, wait - I found the problem. If I create the `results` directory, then it works. Error happens when `$params.outdir` doesn't already exist.

**Comment by pditommaso** [comment]
I see. It raises an exception if the parent path does not exist. It can be improved so that NF creates it if missing.

**Comment by ewels** [comment]
That would be great 👍 

Why is it that the closures don't work? It doesn't really matter that I need to define `params.outdir` twice (pipeline and config file), just curious :)

**Comment by pditommaso** [comment]
Done. Because closure are resolved at runtime by the process execution. Thus are allowed only in the process configuration context. 

The remaining part of the config file only *plain* variables are allowed.

**Comment by pditommaso** [comment]
Fixed with version `0.24.4` .
